Fix #259 and part of #163: Finish state fragment UI structure

BenHenning opened the pull request:
This fixes #259, and part of #163 (core structure to unblock #164, but no tests yet).

This PR introduces the core functionality to enable all supported prototype interactions (minus number with units) and full exploration functionality, per supported peripheral systems for audio and interactions (including answer classification).

In total, this PR includes:
- Introducing a new prototype exploration that supports all prototype interactions (minus numbers with units) that we can expand upon to fully verify the functionality in the exploration player
- A full state adapter that binds each of the interaction types based on ``EphemeralState`` from the ``ExplorationProgressController``
- Generalizing interaction additions in the state fragment to a Dagger module and removal of all other dependencies on interaction IDs
- Replacement of both item selection and state fragment RecyclerView adapters with BindableAdapter (which required some minor changes to BindableAdapter), leading to an end-to-end data-bound UI for the exploration player
- Reworking of navigation buttons to de-couple the Continue interaction so that it can be modeled on its own
- Introduction of automatic answer submission for interactions that do no need an explicit Submit button (like Continue and Multiple Choice), with dynamic support for interactions that can be in either state (ItemSelection)
- Reworking of answer submission to be driven through view models rather than interaction views
- Moved the audio button to a custom toolbar with the audio icon (though it's not currently switching state based on enabled state)
- Updated to a newer (pre-release) AndroidX fragment library to simplify a tricky edge case where we needed to be able to associate a view with a fragment
- Introduced support for displaying feedback cards
- Introduced support for displaying read-only interactions with past answers (and support for converting ``InteractionObject``s to text answers)
- Removed the temporary state name view since that debug view is no longer necessary with hooked up navigation buttons & feedback since folks can observe the current state based on what's displayed
- Introduction of a new view scope for custom views. This scope inherits from the fragment scope, and indirectly from the activity and application scopes. So far this is only used for the selection interaction view, and there's no pattern yet for 'view presenters'.
- Some various bug fixes, including: parsing issues with fractions and reducing the number of individual updates to the recycler view to make the rendering experience less janky
- Minor infrastructure improvements not worth individually calling out (including resolving #259)

Next steps for #163 would be introducing missing tests to guard against regressions. This is blocked on fixing downstream testing issues around swapping dependencies, and fixing existing tests on develop (#295).

Known next steps for #164 include:
- Ensuring the audio button properly switches & animates when the cellular setting is toggled
- Only showing the audio button when voiceovers are available
- Verify that voiceovers play correctly for content and feedback
- Expand the cellular button touch target to be 48dp per accessibility guidelines (it's currently too small and I have some difficulty tapping on it)
- Add animations for RecyclerView updates
- Add a loading state when entering the exploration to remove initial jank
- Improvements in all interactions to improve smoothness (especially item selection's recycler view since it has similar issues to the state fragment recycler view)
- Realtime validation and error reporting for invalid answers in each interaction (where relevant, and including adding min selection support for item selection and reporting the selection count)
- Artificial UIs delays throughout the state fragment to make the conversational flow smoother (e.g. ensuring the 'Continue' button response shows up and is read before continuing to the next state)
- Fixing the enter button crashing the app on the keyboard
- Any other changes that improve the UX of the exploration player, or fix issues
- Ensure the conversational flow looks & behaves correctly in parity with Oppia web (e.g. https://www.oppia.org/explore/0 for reference)

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
numericInputTypeView test case is in progress

nikitamarysolomanpvt committed:
numericInputTypeView test case is in progress

nikitamarysolomanpvt committed:
nit changes as per #223

nikitamarysolomanpvt committed:
nit changes as per #223

veena14cs committed:
working on fetching customization_args

nikitamarysolomanpvt committed:
On configuration changes save and restore the text of numericInputType

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
resolved conflicts

veena14cs committed:
Separated this in its own PR

veena14cs committed:
Delete UrlImageParser.kt

Separated this in its own PR.

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
Update build.gradle

veena14cs committed:
Create HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

nikitamarysolomanpvt committed:
Test Cases for NumericInputInteractionView,
Moved test case to StateFragment

nikitamarysolomanpvt committed:
Test Cases for NumericInputInteractionView,
Moved test case to StateFragment,PR suggestions updated

nikitamarysolomanpvt committed:
PR suggestions updated

nikitamarysolomanpvt committed:
PR suggestions updated

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/build.gradle
#	app/src/main/AndroidManifest.xml
#	app/src/main/res/values/styles.xml

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into text-input-lowfi-numeric-input-interaction-view-part-2

# Conflicts:
#	app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentTest.kt

nikitamarysolomanpvt committed:
nit changes as per PR review

nikitamarysolomanpvt committed:
nit changes as per PR review

nikitamarysolomanpvt committed:
nit changes as per PR review

nikitamarysolomanpvt committed:
nit changes as per PR review

veena14cs committed:
Update HtmlParser.kt

nikitamarysolomanpvt committed:
Number input test Activity test cases

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
setReal type in getPendingAnswer

nikitamarysolomanpvt committed:
updated test cases, disabled long click to prevent pasting value

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'text-input-lowfi-numeric-input-interaction-view-part-2' of https://github.com/oppia/oppia-android into text-input-lowfi-numeric-input-interaction-view-part-2

# Conflicts:
#	app/src/main/AndroidManifest.xml

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
NumericInputInteractionViewTestActivity renamed to InputInteractionViewTestActivity and test case updated

nikitamarysolomanpvt committed:
nit changes

veena14cs committed:
working on testcases

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
added testcases.

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Update InteractionAdapter.kt

veena14cs committed:
Update InteractionAdapter.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
corrected variable names.

veena14cs committed:
updated variables.

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'develop' into introduce-html-parser

veena14cs committed:
Delete InteractionAdapter.kt

veena14cs committed:
deteted unused xml files

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
merged html code.

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Delete UrlImageParserTest.kt

veena14cs committed:
used dagger injection for UrlImageParser.

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
updated tests

veena14cs committed:
updated test

veena14cs committed:
moved to testing package

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
Update HtmlParserTestActivty.kt

veena14cs committed:
fix issues

veena14cs committed:
fix issues.

veena14cs committed:
fixed issues

veena14cs committed:
fixed issues.

veena14cs committed:
Update activity_test_html_parser.xml

veena14cs committed:
fixes

veena14cs committed:
fixes

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
fixed issues.

nikitamarysolomanpvt committed:
nit

veena14cs committed:
fixing issues.

veena14cs committed:
made use of factory

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Update HtmlParserTestActivity.kt

veena14cs committed:
nit

veena14cs committed:
Update AudioFragmentTest.kt

veena14cs committed:
updated

veena14cs committed:
Merge branch 'introduce-image-parser-replica' of https://github.com/oppia/oppia-android into introduce-image-parser-replica

veena14cs committed:
fixes

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Not used in this.

veena14cs committed:
fixes

veena14cs committed:
Merge branch 'introduce-image-parser-replica' of https://github.com/oppia/oppia-android into introduce-image-parser-replica

veena14cs committed:
Merge branch 'introduce-html-parser' into introduce-image-parser-replica

veena14cs committed:
Not used.

veena14cs committed:
Update CellularDataDialogFragmentTest.kt

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
reverted

veena14cs committed:
Update CellularDataDialogFragmentTest.kt

veena14cs committed:
fixes

veena14cs committed:
Update ImageLoader.kt

nikitamarysolomanpvt committed:
NumericInputInteractionView attributes are moved to xml,testcases updated as per PR review of Ben, other nit changes

veena14cs committed:
Update HtmlParser.kt

nikitamarysolomanpvt committed:
NumericInputInteractionView attributes are moved to xml,testcases updated as per PR review of Ben, other nit changes

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
Update build.gradle

veena14cs committed:
fixed issues.

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update ImageParsingAnnotations.kt

nikitamarysolomanpvt committed:
TextInputInteractionView

veena14cs committed:
working on image loader

veena14cs committed:
Injected ImageLoader

veena14cs committed:
updated image loader.

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update HtmlParserTestActivity.kt

nikitamarysolomanpvt committed:
TextInputInteractionView

nikitamarysolomanpvt committed:
test cases

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'lowfi-input-interaction-views' of https://github.com/oppia/oppia-android into lowfi-input-interaction-views

# Conflicts:
#	app/src/main/java/org/oppia/app/activity/InputInteractionViewTestActivity.kt
#	app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt
#	app/src/main/res/layout/activity_numeric_input_interaction_view_test.xml
#	app/src/main/res/values/dimens.xml
#	app/src/sharedTest/java/org/oppia/app/activity/InputInteractionViewTestActivityTest.kt

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'lowfi-input-interaction-views' of https://github.com/oppia/oppia-android into lowfi-input-interaction-views

# Conflicts:
#	app/src/main/java/org/oppia/app/activity/InputInteractionViewTestActivity.kt
#	app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt
#	app/src/main/res/layout/activity_numeric_input_interaction_view_test.xml
#	app/src/main/res/values/dimens.xml
#	app/src/sharedTest/java/org/oppia/app/activity/InputInteractionViewTestActivityTest.kt

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'lowfi-input-interaction-views' of https://github.com/oppia/oppia-android into lowfi-input-interaction-views

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'text-input-lowfi-numeric-input-interaction-view-part-2' of https://github.com/oppia/oppia-android into text-input-lowfi-numeric-input-interaction-view-part-2

# Conflicts:
#	app/src/sharedTest/java/org/oppia/app/activity/InputInteractionViewTestActivityTest.kt

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
NumberWithUnitsInputInteractionView

veena14cs committed:
Merge branch 'develop' into introduce-image-parser-replica

veena14cs committed:
Update GlideImageLoader.kt

veena14cs committed:
removed annotation

veena14cs committed:
Update HtmlParser.kt

veena14cs committed:
line break

veena14cs committed:
fix

veena14cs committed:
fix comments

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
injected context

veena14cs committed:
working on testcase.

veena14cs committed:
testcase added

nikitamarysolomanpvt committed:
NumberWithUnitsInputInteractionView

veena14cs committed:
added test case

veena14cs committed:
testcases

veena14cs committed:
Update build.gradle

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
fixes

nikitamarysolomanpvt committed:
Test cases Number With Units

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into lowfi-number-with-units-input-interaction-view

veena14cs committed:
Update UrlImageParserTest.kt

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

veena14cs committed:
changed to CutomTarget

nikitamarysolomanpvt committed:
nit

BenHenning committed:
Merge branch 'introduce-image-parser-replica' into temp-integration-branch-for-state

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt

BenHenning committed:
Merge branch 'lowfi-number-with-units-input-interaction-view' into temp-integration-branch-for-state

BenHenning committed:
Attempt to rebase content-card onto develop.

BenHenning committed:
Merge branch 'introduce-image-parser-replica' into content-card-from-develop

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt

BenHenning committed:
Merge branch 'develop' into introduce-image-parser-replica-up-to-date

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt

BenHenning committed:
Merge branch 'introduce-image-parser-replica-up-to-date' into content-card-from-develop

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt

BenHenning committed:
Manually apply rejections and remove rej files.

BenHenning committed:
Add missing pieces implied in merge-fix that are needed for this branch.

BenHenning committed:
Merge branch 'content-card-from-develop' into temp-integration-branch-for-state

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt

BenHenning committed:
Maunally apply multiple-single-input-interaction changes to develop
branch (with patch rejections).

BenHenning committed:
Merge branch 'content-card-from-develop' into multiple-single-input-interaction-from-develop

Conflicts:
    app/src/main/java/org/oppia/app/player/state/StateAdapter.kt

BenHenning committed:
Manually apply rejected diffs.

BenHenning committed:
Fix broken build after merges.

BenHenning committed:
Add missing changes from merge-fix that are needed to enable state
content binding.

BenHenning committed:
Merge branch 'content-card-from-develop' into multiple-single-input-interaction-from-develop

Conflicts:
	app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

BenHenning committed:
Resolve unresolved conflicts.

BenHenning committed:
Merge branch 'content-card-from-develop' into temp-integration-branch-for-state

BenHenning committed:
Merge branch 'multiple-single-input-interaction-from-develop' into temp-integration-branch-for-state

BenHenning committed:
Make an exploration with all prototype interactions work end-to-end.

This includes:
- Introducing an exploration with all prototype interactions
- Introducing support for binding all interaction types
- Fixing some parsing issues with fractions
- Rewiring the interaction object creation for views to be generic
- Removing temporary answer handling everywhere
- Other random fixes

This includes no tests, and there are a few other fixes still necessary
for the core learning experience to fully work.

BenHenning committed:
Add support for displaying feedback items.

BenHenning committed:
Add support for reshowing submitted answers.

BenHenning committed:
Make 'Continue' button a separate interaction (but keep the existing
general 'Continue' navigation button). Rework how navigation buttons
behave. Introduce automatic answer submission for some interactions
(Continue, ItemSelection if it shows radio buttons, and MultipleChoice).

BenHenning committed:
Move audio functionality to a custom toolbar, remove the dummy audio
button, and remove the transient state name at the top of the state
fragment.

BenHenning committed:
Simplify the presenter by updating ExplorationProgressController to feed
information about the next and current state to allow the state presenter
to only focus on rendering the current state.

jamesxu0 commented:
@BenHenning Sean wanted me to let you know that creating the AudioFragment currently crashes StateFragment because currentStateName isn't set anywhere.

BenHenning committed:
First batch of work corresponding to moving the entirety of state
fragment's recycler view (and child views/recycler view) to be bound via
data-binding.

BenHenning committed:
Finish updating item selection recycler view to use data binding. This
required introducing a ViewComponent that inherits from FragmentComponent
in the Dagger dependency graph. This also involved updating the app to use
a newer AndroidX fragment library that's still pre-release in order to
gain access to the necessary functionality to associate views with their
fragments.

The two custom adapters have been removed in this change.

BenHenning committed:
Some minor fixes to polish up the experience a bit and reduce jank.

BenHenning committed:
Fix broken voiceover audio button.

BenHenning commented:
Thanks @jamesxu0 --I must have broken it during one aspect of my refactor. The latest commit fixes it.

BenHenning committed:
Address one TODO to generalize adding interactions by leveraging a Dagger
module.

BenHenning commented:
I believe this work is now in a relatively finalized state, minus tests. I'm now waiting on downstream work to make it into develop so that I can prep this for review. Some TODOs may still need resolving, but most will be filed as long-term issues to track.

seanlip commented:
@BenHenning -- please could you make clearer which downstream work exactly needs to be completed in order to unblock this PR?

Thanks!

BenHenning committed:
Merge branch 'develop' into finish-state-fragment-ui-structure

Conflicts:
	app/build.gradle
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/java/org/oppia/app/testing/HtmlParserTestActivity.kt
	app/src/main/java/org/oppia/app/topic/play/TopicPlayFragment.kt
	app/src/main/res/values/strings.xml
	app/src/sharedTest/java/org/oppia/app/parser/HtmlParserTest.kt
	utility/src/main/java/org/oppia/util/parser/GlideImageLoaderModule.kt
	utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt

BenHenning committed:
Fix broken test.

BenHenning committed:
Fix broken domain tests by allowing feedback to be either a string or an
object, and loading about_oppia instead of oppia_exploration since the
former is expected by ExplorationProgressControllerTest.

BenHenning committed:
Fix broken app test builds (tests still fail, but at least run now).

BenHenning commented:
Blocking PRs: #275 and #276. Also technically #263 but that's on hold.

seanlip commented:
Yep, let's actually consider #263 non-blocking. We can bring it in later at a future date.

BenHenning commented:
I need to bring this up to develop once #276 is checked in, and then remove the number with input support from both the player and the test exploration.

BenHenning committed:
Add TODO to fix feedback parsing hack.

BenHenning committed:
Remove TODOs in StringToFractionParser.

BenHenning committed:
Resolve open TODOs for the state portion of the player, or associate
them with issues.

BenHenning committed:
Associate 2 more TODOs with issues.

BenHenning committed:
Add tests for new next state behavior in ExplorationProgressController,
and resolve last unattributed TODO.

BenHenning committed:
Remove number with units support.

BenHenning committed:
Merge branch 'develop' into finish-state-fragment-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/activity/ActivityComponent.kt
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/StateAdapter.kt
	app/src/main/java/org/oppia/app/player/state/StateFragment.kt
	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/ContentViewModel.kt
	app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionContentViewModel.kt
	app/src/main/res/layout/content_item.xml
	app/src/main/res/layout/home_fragment.xml
	app/src/main/res/layout/item_selection_interaction_items.xml
	app/src/main/res/layout/multiple_choice_interaction_items.xml
	app/src/main/res/layout/selection_interaction_item.xml
	app/src/main/res/layout/state_fragment.xml
	app/src/main/res/values/strings.xml
	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
	domain/src/main/java/org/oppia/domain/util/StateRetriever.kt
	utility/src/main/java/org/oppia/util/parser/ExplorationHtmlParserEntityType.kt
	utility/src/main/java/org/oppia/util/parser/HtmlParserEntityTypeModule.kt

BenHenning committed:
Post-merge fixes and clean-ups.

BenHenning committed:
Merge branch 'develop' into finish-state-fragment-ui-structure

Conflicts:
	app/src/main/AndroidManifest.xml

BenHenning committed:
Post-merge fix.

BenHenning commented:
NB: If you want to test this change with the latest home fragment changes, move the play exploration button in home_fragment.xml to be above the recycler view and set its visibility to visible. It should load the prototype exploration and play as expected.

BenHenning commented:
Folks--this is ready for review. Sorry for the really large PR, but trying to get a working integration requires many changes, and we're a bit strapped for time so splitting this up is difficult. Please focus on specifically broken or confusing aspects. If you have feedback for design, please follow up with a top-level comment and I'll file an issue to track changing it.

There are no new tests introduced here, and some existing tests may break due to these changes. That's working as intended since my intention is to unblock downstream final UI work, then turn back to finishing up all necessary core tests for this experience (since there's both investigation and infrastructure work needed to properly test the core exploration experience). Please let me know if you have any specific concerns with this.

This is a long PR, so please do read through it. I likely made mistakes, and I'm unlikely to find them without your help.

veena14cs commented on app/src/main/java/org/oppia/app/activity/ActivityComponent.kt:
```diff
@@ -38,7 +39,8 @@ interface ActivityComponent {
   fun inject(explorationActivity: ExplorationActivity)
   fun inject(homeActivity: HomeActivity)
-  fun inject(htmlParserTestActivity: HtmlParserTestActivity)
+  fun inject(htmlParserTestActivty: HtmlParserTestActivity)
```
Typo: in word `Activty`.

rt4914 replied:
+1

BenHenning replied:
Fixed. Nice catch!

veena14cs commented on app/src/main/java/org/oppia/app/customview/interaction/TextInputInteractionView.kt:
```diff
@@ -16,13 +16,4 @@ class TextInputInteractionView @JvmOverloads constructor(
-    return interactionObjectBuilder.build()
-  }
-}
+) : EditText(context, attrs, defStyle)
```
This file has got unused imports.

BenHenning replied:
Removed.

veena14cs commented on app/src/main/res/layout/fraction_interaction_item.xml:
```diff
@@ -0,0 +1,39 @@
+      android:padding="8dp"
+      app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintTop_toTopOf="parent">
+      <org.oppia.app.customview.interaction.FractionInputInteractionView
```
This view needs to be restricted with digits `android:digits="0123456789/- "`. If we enter any string other than number it crashes.

BenHenning replied:
Oh! Interesting. For some reason when I tried this before, it didn't seem to filter out the extra characters but it's working now. Re-added--thanks for pointing this out to me.

veena14cs commented on app/src/main/res/layout/exploration_activity.xml:
```diff
@@ -1,8 +1,43 @@
+    android:id="@+id/exploration_fragment_placeholder"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    app:layout_behavior="@string/appbar_scrolling_view_behavior" />
```
`appbar_scrolling_view_behavior` doesn't seem to be present in strings.xml file.

BenHenning replied:
Good catch--it's defined in a strings.xml file we pull in from the material library. It should be ctrl + clickable in Android Studio once you sync Gradle & rebuild. I don't quite like how they implemented this mechanism since it's confusing to reference a string from outside the project, but it's correct and works.

veena14cs commented on app/src/main/java/org/oppia/app/player/state/StateViewModel.kt:
```diff
@@ -1,17 +1,39 @@
+import androidx.databinding.ObservableArrayList
 import androidx.databinding.ObservableField
+import androidx.databinding.ObservableList
+import androidx.databinding.ObservableMap
```
This file has unused imports.

rt4914 replied:
+1

BenHenning replied:
I'm starting to see a pattern emerge. :) I ran optimize imports on the whole project and verified that the changes it outputted are reasonable. It did pull in a lot of extra files, but all the changes seem reasonable and trivial.

veena14cs commented on app/src/main/java/org/oppia/app/testing/UrlImageParserTestActivity.kt:
```diff
@@ -0,0 +1,17 @@
+import org.oppia.app.activity.InjectableAppCompatActivity
+
+/** This is a dummy activity to test [UrlImageParser]. */
+class UrlImageParserTestActivity : InjectableAppCompatActivity() {
```
This test activity is not added in manifest or it is unused.

BenHenning replied:
Good catch, removed this & the corresponding test. A likely artifact from when I was merging in the pre-finalized upstream changes.

veena14cs commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentContentCardTest.kt:
```diff
@@ -0,0 +1,31 @@
+// TODO(#205): Add test case for image parsing once PR #205 is merged.
+/** Tests for [VIEW_TYPE_CONTENT]. */
+@RunWith(AndroidJUnit4::class)
+class StateFragmentContentCardTest {
```
Test case from this file is moved to `StateFragmentTest`, hence you can delete this file.

BenHenning replied:
Sounds good--thanks! File deleted.

veena14cs commented on app/src/sharedTest/java/org/oppia/app/player/state/StateSelectionInteractionTest.kt:
```diff
@@ -0,0 +1,120 @@
+
+/** Tests for SelectionInteraction [MULTIPLE_CHOICE_INPUT], [ITEM_SELECT_INPUT]. */
+@RunWith(AndroidJUnit4::class)
+class StateSelectionInteractionTest {
```
Test case from this file is moved to `StateFragmentTest`, hence you can delete this file.

BenHenning replied:
Done.

veena14cs commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -24,9 +24,8 @@ class UrlImageParser private constructor(
   private val entityType: String,
   private val entityId: String,
-  private val imageLoader:ImageLoader
+  @ImageLoaderAnnotation private val imageLoader: ImageLoader
```
This class is not up-to-date with develop. Implementation of `UrlImageParser` is changed it doesn't include `ImageLoaderAnnotation`.

veena14cs replied:
Ditto: testcases for `UrlImageParser` too.

BenHenning replied:
Good catch. I think I merged this part wrong when I was pulling in the newer changes, fixed both (removed the annotation and UrlImageParser sets).

veena14cs reviewed (changes requested):
Thanks  @BenHenning for this integration and major changes in interactions. It is really lengthy PR.
There are some nit changes and some needs to be updated with develop branch.

rt4914 commented:
@BenHenning 

<img width="345" alt="Screenshot 2019-11-05 at 9 35 59 PM" src="https://user-images.githubusercontent.com/9396084/68224312-5c883e00-0014-11ea-9d90-a8951e59a220.png">

Check this screenshot, in ItemSelection there are two issues:

1. Let's assume that maximum item selection count is 3 and if the user tries to select the 4 options, then it does not get selected which is correct but here user should get a feedback in the form of toast or warning, that they cannot select more than 3 options.
2. Once correct answer is submitted, the "Continue" button is directly visible without any feedback to user that it is a correct answer.
3. 
<img width="333" alt="Screenshot 2019-11-05 at 9 43 11 PM" src="https://user-images.githubusercontent.com/9396084/68224787-4d55c000-0015-11ea-81b0-bbd662350bc4.png">

**Most Important UX Issue**
If the learner has selected some incorrect answer in ItemSelection or MultipleChoiceInput, I think radio buttons and checkboxes should be removed in that case which would have two advantages:
a. The content-length on the screen will be less which will lead to less scrolling back to the question on top.
b. Learner won't get confused between already selected radio/checkbox (in which they cannot edit) and the final radio/checkbox (in which they can edit.) 




Other issues:
Audio Player is not working correctly, this might be because of the current exploration but I would still think it would be nice to if we can introduce dummy voiceover to test this functionality.

Learner should get a feedback on correct answer:
<img width="323" alt="Screenshot 2019-11-05 at 9 41 44 PM" src="https://user-images.githubusercontent.com/9396084/68224685-17184080-0015-11ea-8f2f-7c0d21fd30ab.png">


Those some of these are related to high-fi, it is worth mentioning it here too.

rt4914 commented on app/build.gradle:
```diff
@@ -62,15 +62,19 @@ dependencies {
       'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.1',
       'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.2.1',
+      'org.mockito:mockito-core:2.7.22',
+      'com.github.bumptech.glide:glide:4.9.0',
```
This library has already been imported above.

BenHenning replied:
Good catch. Removed.

rt4914 commented on app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt:
```diff
@@ -17,13 +17,4 @@ class NumericInputInteractionView @JvmOverloads constructor(
-    return interactionObjectBuilder.build()
-  }
-}
+) : EditText(context, attrs, defStyle)
```
Remove unused imports.

BenHenning replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/fragment/InjectableFragment.kt:
```diff
@@ -16,8 +21,16 @@ abstract class InjectableFragment: Fragment() {
+    return super.onCreateView(inflater, container, savedInstanceState)
+  }
+
+  fun createViewComponent(view: View): ViewComponent {
```
This function is not getting used.

BenHenning replied:
Ah, I forget why I added the view component. I think it was for the item selection view--it's a nice idea, but we don't need it right now. Removed the whole thing.

BenHenning replied:
Actually I made a mistake--this is used by SelectionInteractionView since it's using a new view scope. I also realized that I forgot to document this  in the top-level PR and will make that adjustment. The method appears used in Android Studio once the project is built.

rt4914 replied:
sounds good.

rt4914 commented on app/src/main/java/org/oppia/app/fragment/InjectableFragment.kt:
```diff
@@ -16,8 +21,16 @@ abstract class InjectableFragment: Fragment() {
     fragmentComponent = (requireActivity() as InjectableAppCompatActivity).createFragmentComponent(this)
   }
+
+  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
```
Redundant override method

BenHenning replied:
Removed. I think this was an artifact from when I was trying to create a custom view component.

rt4914 replied:
ack

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationFragmentPresenter.kt:
```diff
@@ -31,6 +31,10 @@ class ExplorationFragmentPresenter @Inject constructor(
     return binding
   }
 
+  fun handlePlayAudio() {
```
There are unused imports in this file

BenHenning replied:
Cleaned up.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/answerhandling/InteractionAnswerHandler.kt:
```diff
@@ -0,0 +1,27 @@
+}
+
+/**
+ * A callback that will be called by [InteractionAnswerPusher]s when a user submits an answer. To be implemented by
```
What is this `InteractionAnswerPusher` because I am not able to find any such file?

BenHenning replied:
An old interface that has been merged with the interaction answer retriever to become ``InteractionAnswerHandler``. Good catch--updated the documentation.

rt4914 replied:
ack

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/ContinueInteractionViewModel.kt:
```diff
@@ -0,0 +1,28 @@
+class ContinueInteractionViewModel(
+  private val interactionAnswerReceiver: InteractionAnswerReceiver, existingAnswer: InteractionObject?,
+  val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
```
Nit: `) : StateItemViewModel()` 

Spacing issue

BenHenning replied:
Hmm, we're really not consistent about this one. I'm inclined to leave it as-is since I noticed most of the other classes in this package follow this style. I haven't been enforcing it personally, either, since it's not super consistent for classes. For methods I've been more pedantic about it.

It's worth coming up with a style rule for this and enforcing it once we have linting and auto-formatting set up the way we like.

BenHenning replied:
Also there appear to be close to 100 occurrences of this in the codebase, though getting a good Regex pattern to find + replace them all is a bit tricky so that's probably an overestimate.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/FeedbackViewModel.kt:
```diff
@@ -0,0 +1,4 @@
+package org.oppia.app.player.state.itemviewmodel
+
+/** [ViewModel] for feedback blurbs. */
+class FeedbackViewModel(val htmlContent: CharSequence): StateItemViewModel()
```
Nit: `) : StateItemViewModel()`

BenHenning replied:
See earlier reply. Going to resolve this & later similar comments without follow-ups.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/FractionInteractionViewModel.kt:
```diff
@@ -0,0 +1,20 @@
+
+class FractionInteractionViewModel(
+  existingAnswer: InteractionObject?, val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
```
nit: `) : StateItemViewModel()`

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/InteractionViewModelModule.kt:
```diff
@@ -0,0 +1,56 @@
+
+/** Module to provide interaction view model-specific dependencies. */
+@Module
+class InteractionViewModelModule {
```
Should we introduce `EndExploration` Id in this ?

BenHenning replied:
Not now, I think. It would only be needed if we needed to display something for the end exploration, but we don't as of yet. I'll make that a bit clearer in the documentation for this module, too (ie, not all interactions necessarily need to be represented here).

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/NumericInputViewModel.kt:
```diff
@@ -0,0 +1,19 @@
+
+class NumericInputViewModel(
+  existingAnswer: InteractionObject?, val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
```
Nit: `) : StateItemViewModel()`

spacing issue

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionContentViewModel.kt:
```diff
@@ -1,9 +1,20 @@
+class SelectionInteractionContentViewModel(
+  val htmlContent: String, private val itemIndex: Int, isAnswerInitiallySelected: Boolean, val isReadOnly: Boolean,
+  private val selectionInteractionViewModel: SelectionInteractionViewModel
+): ObservableViewModel() {
```
Nit: `) : ObservableViewModel() `

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+class SelectionInteractionViewModel(
+  val explorationId: String, interaction: Interaction, private val interactionAnswerReceiver: InteractionAnswerReceiver,
+  existingAnswer: InteractionObject?, val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
```
Nit: `) : StateItemViewModel()`

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+  val explorationId: String, interaction: Interaction, private val interactionAnswerReceiver: InteractionAnswerReceiver,
+  existingAnswer: InteractionObject?, val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
+  val interactionId: String = interaction.id
```
This can be made `private`

BenHenning replied:
Nice catch for these visibility reductions. Done here & elsewhere in this file.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+    // when either of the counts are not specified.
+    interaction.customizationArgsMap["maxAllowableSelectionCount"]?.signedInt ?: minAllowableSelectionCount
+  }
+  val selectedItems = computeSelectedItems(
```
This can be made `private`

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/SelectionInteractionViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+import org.oppia.app.player.state.answerhandling.InteractionAnswerReceiver
+
+/** ViewModel for multiple or item-selection input choice list. */
+class SelectionInteractionViewModel(
```
There is another file in this folder namely, `SelectionInteractionCustomizationArgsViewModel`, which is not getting used anywhere, maybe you can remove that file in this PR.

BenHenning replied:
Good catch! Removed.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/StateItemViewModel.kt:
```diff
@@ -0,0 +1,6 @@
+import org.oppia.app.viewmodel.ObservableViewModel
+
+/** The root [ViewModel] for all individual items that may be displayed in the state fragment recycler view. */
+abstract class StateItemViewModel: ObservableViewModel()
```
Nit: `abstract class StateItemViewModel : ObservableViewModel()`

rt4914 commented on app/src/main/java/org/oppia/app/player/state/itemviewmodel/TextInputViewModel.kt:
```diff
@@ -0,0 +1,19 @@
+
+class TextInputViewModel(
+  existingAnswer: InteractionObject?, val isReadOnly: Boolean
+): StateItemViewModel(), InteractionAnswerHandler {
```
Nit: `) : StateItemViewModel()`

rt4914 commented on app/src/main/res/layout/content_item.xml:
```diff
@@ -1,10 +1,9 @@
     <variable
       name="htmlContent"
-      type="String" />
+      type="CharSequence"/>
```
Nit: space before closing tag.

BenHenning replied:
Per this & later comments, I reformatted all of the layout files in the project. It again pulled in some extra changes, but they seem reasonable and this PR at this point is becoming a collection of random refactors, anyway.

Consider this a resolution of later similar comments (I won't reply to them, but they should be addressed with this blanket clean-up).

rt4914 replied:
Makes sense.

rt4914 commented on app/src/main/res/layout/content_item.xml:
```diff
@@ -22,7 +21,8 @@
+      android:visibility="@{htmlContent.length() > 0 ? View.VISIBLE : View.GONE, default=gone}"
       android:textColor="@color/oppiaPrimaryText"
-      android:textSize="16sp" />
+      android:textSize="16sp"/>
```
nit: space before closing tag

rt4914 commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -0,0 +1,43 @@
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+  xmlns:app="http://schemas.android.com/apk/res-auto">
+  <data>
+    <import type="android.view.View"/>
```
Nit: space before closing tag, everywhere in this file.

rt4914 commented on app/src/main/res/layout/feedback_item.xml:
```diff
@@ -0,0 +1,23 @@
+      android:padding="12dp"
+      android:text="@{htmlContent}"
+      android:textColor="@color/oppiaPrimaryText"
+      android:textSize="16sp"/>
```
Nit: space before closing tag here and above in this file.

rt4914 commented on app/src/main/res/layout/fraction_interaction_item.xml:
```diff
@@ -0,0 +1,39 @@
+        android:padding="8dp"
+        android:singleLine="true"
+        android:enabled="@{!viewModel.isReadOnly}"
+        android:text="@={viewModel.answerText}"/>
```
Nit: space before closing tag.

rt4914 commented on app/src/main/res/layout/item_selection_interaction_items.xml:
```diff
@@ -1,36 +1,36 @@
-      android:text="@{viewModel.htmlContent}"
-      android:textColor="@color/oppiaDarkBlue" />
+      android:text="@{htmlContent}"
+      android:textColor="@color/oppiaDarkBlue"/>
```
Nit: space before closing tag.

rt4914 commented on app/src/main/res/layout/multiple_choice_interaction_items.xml:
```diff
@@ -1,35 +1,35 @@
       android:padding="4dp"
       android:text="@{htmlContent}"
-      android:textColor="@color/oppiaDarkBlue" />
+      android:textColor="@color/oppiaDarkBlue"/>
```
Nit: Closing tag space issue.

rt4914 commented on app/src/main/res/layout/numeric_input_interaction_item.xml:
```diff
@@ -0,0 +1,32 @@
+  <data>
+    <variable
+      name="viewModel"
+      type="org.oppia.app.player.state.itemviewmodel.NumericInputViewModel"/>
```
Nit: Space before closing tag.

rt4914 commented on app/src/main/res/layout/selection_interaction_item.xml:
```diff
@@ -3,6 +3,12 @@
+  <data>
+    <variable
+      name="viewModel"
+      type="org.oppia.app.player.state.itemviewmodel.SelectionInteractionViewModel"/>
```
Nit: space before closing tag.

rt4914 commented on app/src/main/res/layout/state_button_item.xml:
```diff
@@ -5,7 +5,7 @@
     <variable
       name="buttonViewModel"
-      type="org.oppia.app.player.state.itemviewmodel.StateButtonViewModel"/>
+      type="org.oppia.app.player.state.itemviewmodel.StateNavigationButtonViewModel"/>
```
Nit: space before closing tag.

rt4914 commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -49,10 +19,11 @@
       app:layout_constraintEnd_toEndOf="parent"
       app:layout_constraintStart_toStartOf="parent"
-      app:layout_constraintTop_toBottomOf="@+id/dummy_state_name"/>
+      app:layout_constraintTop_toTopOf="parent"/>
```
Nit: space before closing tag

rt4914 commented on app/src/main/res/layout/text_input_interaction_item.xml:
```diff
@@ -0,0 +1,31 @@
+  <data>
+    <variable
+      name="viewModel"
+      type="org.oppia.app.player.state.itemviewmodel.TextInputViewModel"/>
```
Nit: space before closing tag.

rt4914 reviewed (approved):
Really great implementation. Mostly nit changes and also, I have addressed some of the issues which I think needs to be fixed in this PR or in High-fi.

rt4914 reviewed (approved):
Really great implementation. Mostly nit changes and also, I have addressed some of the issues which I think needs to be fixed in this PR or in High-fi.

BenHenning commented:
Thanks @veena14cs & @rt4914 -- I will address comments tomorrow since it's getting a bit late for me.

Rajat, I think most of that should be resolved as part of highfi work, but I agree with hiding the item selection & radio buttons on wrong answers. Let's talk about this during the Android team meeting later. I need a refresher on how Oppia web handles this, too.

Also note that the item selection interaction in the exploration itself has a bug--it doesn't display feedback for the correct answer. I think you might have run into this--it's not an issue with the implementation, but the exploration itself. I'll make a note to patch that as part of resolving the open comments.

jamesxu0 reviewed (approved):
LGTM other than the unresolved comments.

BenHenning committed:
Merge branch 'develop' into finish-state-fragment-ui-structure

Conflicts:
	app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt
	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
	app/src/sharedTest/java/org/oppia/app/testing/InputInteractionViewTestActivityTest.kt
	domain/src/main/java/org/oppia/domain/util/StateRetriever.kt

BenHenning committed:
Remove number with units tests.

BenHenning committed:
First round of addressing reviewer comments.

BenHenning committed:
Remove view scope & component.

BenHenning committed:
Second round of addressing reviewer comments. Includes a project-wide
optimization of all imports.

BenHenning committed:
Round 3 of addressing reviewer changes: removed UrlImageParserTest since
it was pulled in with earlier, unfinalized changes and we have since
realized it's not possible to test the image parser based on the current
project setup.

BenHenning committed:
Round 4 of addressing reviewer comments: reformat all layout XML files.

BenHenning committed:
Round 5 of addressing reviewer comments.

BenHenning committed:
Revert "Remove view scope & component."

This reverts commit 99783d53bae9d4233ef5c2de96f17a1a2499ecd8.

BenHenning committed:
Post-clean up fixes and other minor adjustments.

BenHenning committed:
Ensure ExplorationActivityTest passes even though it doesn't have
interesting tests yet.

BenHenning committed:
Temporary workaround to make the prototype exploration accessible with
recent topic & home fragment changes.

Also, hide the topic button in the home fragment since the topic is
accessible directly via the home fragment tiles.

BenHenning committed:
Round 6 of addressing review comments.

BenHenning committed:
Update prototype exploration to provide feedback after the multi-item
item selection state for the correct answer so that the player doesn't
seem broken.

BenHenning commented:
Thanks everyone for looking at this. There were a large number of style & small fixes in the latest batch, along with a few minor functional changes & build fixes. See the commit log for more detail.

@rt4914 I fixed the prototype exploration having no output which led to the player feeling broken. I did not address your other comments around the multiple item selection & multiple choice blocks showing up, or the correct answer box. We'll need to address both in highfi work (I don't want to bloat this PR anymore).

Also note that for the correct answer box, that's based on whether an answer is labeled as correct. Most answers aren't labeled as correct in the explorations saved in Oppia's backend, so we'll often not show the box even when it's implemented. We should still implement it though, especially for questions (@jamesxu0 FYI). Please let me know if you want to discuss this further.

rt4914 commented:
@BenHenning  I have taken overview look and also checked all the comments. Almost all replies makes sense and most of my questions will get answered in High-fi work.

veena14cs reviewed (approved):
Looks great.

BenHenning commented:
One follow up from talking to @seanlip: we should be showing the completed answers for each interaction, rather than the whole interaction. This works better with the conversational dialog, and it matches Oppia web: https://www.oppia.org/explore/0. Updated my top comment to include this as a future work item.

This might make more sense to do as part of follow-up #163 work since it's a bit more structural. I'll note that in #163 as well.
